pkg/virt-handler/vm.go

```diff
@@ -187,3 +203,224 @@ func MapPersistentVolumes(vm *v1.VM, restClient cache.Getter, namespace string)
+
+					// Set the profileID according to whether Kubernetes policy is required.
+					// If it's not, then just use the network name (which is the normal behavior)
+					// otherwise use one based on the Kubernetes pod's Namespace.
+					if cniConf.Policy.PolicyType == "k8s" {
+						endpoint.Spec.Profiles = []string{fmt.Sprintf("k8s_ns.%s", namespace)}
+					} else {
+						endpoint.Spec.Profiles = []string{cniConf.Name}
+					}
+
+					// Populate the endpoint with the output from the IPAM plugin.
+					if err = PopulateEndpointNets(endpoint, result); err != nil {
+						// Cleanup IP allocation and return the error.
+						ReleaseIPAllocation(logger, cniConf.IPAM.Type, conf.Bytes)
+						return vm, err
+					}
+					logging.DefaultLogger().Object(vm).Info().Msgf("Populated endpoint: %v", endpoint)
+
+					// create a tap device
+					nameTemplate := fmt.Sprintf("tap-%s-%%d", vmCopy.Spec.Domain.UUID[0:8])
+					ifName, err := tuntap.CreatePersistentIface(nameTemplate, tuntap.Tap)
+					if err != nil {
+						return vm, fmt.Errorf("Failed to create a tap device: %v", err)
+					}
+
+					link, err := netlink.LinkByName(ifName)
+					if err != nil {
+						return vm, fmt.Errorf("cannot find link %q", ifName)
+					}
+					logging.DefaultLogger().Object(vm).Info().Msgf("Link(tap) %v created", link)
+
+					if err = netlink.LinkSetUp(link); err != nil {
+						return nil, fmt.Errorf("cannot set link up %q", ifName)
+					}
+
+					mac := link.Attrs().HardwareAddr
+					endpoint.Spec.MAC = &cnet.MAC{HardwareAddr: mac}
+					endpoint.Spec.InterfaceName = ifName
+					logger.WithField("endpoint", endpoint).Info("Added Mac and interface name to endpoint")
+
+					// Write the endpoint object (either the newly created one, or the updated one)
+					if _, err := calicoClient.WorkloadEndpoints().Apply(endpoint); err != nil {
+						// Cleanup IP allocation and return the error.
+						ReleaseIPAllocation(logger, cniConf.IPAM.Type, conf.Bytes)
+						return vm, err
+					}
+					logger.Info("Wrote updated endpoint to datastore")
+
+					logging.DefaultLogger().Object(vm).Info().Msgf("Mapping %s", ifName)
+					newIface := v1.Interface{}
+
+					newIface.Type = "ethernet"
+					newIface.MAC = new(v1.MAC)
+					newIface.MAC.MAC = mac.String()
+					newIface.Target = new(v1.InterfaceTarget)
+					newIface.Target.Device = ifName
+					newIface.Model = new(v1.Model)
+					newIface.Model.Type = "virtio"
+
+					vmCopy.Spec.Domain.Devices.Interfaces[idx] = newIface
+					return vmCopy, nil
+				}
+			}
+			return vm, fmt.Errorf("No %s network provider found in %s", iface.Source.Network, pluginDir)
+		}
+	}
+	return vmCopy, nil
+}
+
+func DeleteInterfaces(vm *v1.VM, namespace string) error {
+	netConfDir := os.Getenv("CNI_CONF")
+	files, err := libcni.ConfFiles(netConfDir)
+	switch {
+	case err != nil:
+		return err
+	case len(files) == 0:
+		return fmt.Errorf("No networks found in %s", netConfDir)
+	}
+	sort.Strings(files)
+	for _, file := range files {
```

Comment:
If you can use the direct attach, I think you don't even have to ever clean up ... The macvtaps would be removed by libvirt ...